### PR TITLE
CBG-1436: Switch sharded import DCP feed to gocb v2

### PR DESCRIPTION
**Pull request by markspolakovs**
CBG-1436

Switches the import DCP feed to GoCB v2.

**NB**: This will only be used for EE - CE will likely be addressed as part of CBG-1143 as it uses the same code.

# Dependencies

- [x] cbgt changes (https://review.couchbase.org/c/cbgt/+/177563, https://review.couchbase.org/c/cbgt/+/177568/2, https://review.couchbase.org/c/cbgt/+/177572/2)
- [x] Bump cbgt in go.mod once those are merged

## [Integration Tests](https://jenkins.sgwdev.com/job/SyncGateway-Integration/build?delay=0sec)
- [x] `server=7.0.3,protocol=couchbase` https://jenkins.sgwdev.com/job/SyncGateway-Integration/465/
- [x] `server=6.6.5,protocol=couchbase` https://jenkins.sgwdev.com/job/SyncGateway-Integration/466/

### CODE REVIEW

File: base/bucket.go
Comment (markspolakovs) [inline base/bucket.go]: Can be reverted once https://review.couchbase.org/c/cbgt/+/177568/2 goes in.

File: base/dcp_sharded.go
Comment (markspolakovs) [inline base/dcp_sharded.go:312]: Needs https://review.couchbase.org/c/cbgt/+/177563 to go in first

File: base/bucket.go
Comment (adamcfraser) [inline base/bucket.go]: Instead of passing a dcp bool to this function, I think it would be better to pass an options struct with bool flags for values we want to be configurable (kv pool size now, and I anticipate the memcached read cache buffer in the near future).

File: base/bucket.go
Comment (adamcfraser) [inline base/bucket.go]: Do we actually want to update the spec here?  If that's the case, can we skip the process described above and just define a custom bucket spec for use by DCP?

File: base/dcp_feed_type.go
Comment (adamcfraser) [inline base/dcp_feed_type.go:245]: If any of these handlers have identical behaviour regardless of DCP feed type, we can just use the shared definition.  Is that the case here?

File: base/dcp_sharded.go
Comment (adamcfraser) [inline base/dcp_sharded.go:92]: I know it's not part of this ticket, but would like to be sure that we're tracking the upgrade work when moving from a gocouchbase to a gocb feed.  Feed definitions end up stored in the bucket and have an association with checkpoints - need to validate that we're resuming from the previous feed's checkpoints when upgrading.

File: base/dcp_test.go
Comment (adamcfraser) [inline base/dcp_test.go]: Should we be keeping the test for the old approach around, and only run with older CBS versions?  Even as a disabled, dev-only test.

Comment (adamcfraser) [reply base/dcp_test.go]: This is maybe a more general question about our integration testing - with all the changes for collections, I suspect we should think about having an option to run the integration tests against a 6.5 server to catch regressions.

**Review by adamcfraser — COMMENTED** [review]
General approach looks good, some comments on the details.

File: base/dcp_sharded.go
Comment (markspolakovs) [reply base/dcp_sharded.go:92]: Ack, filed https://issues.couchbase.com/browse/CBG-2213

File: base/bucket.go
Comment (markspolakovs) [reply base/bucket.go]: Not quite sure I follow - the doc comment on `KvPoolSize` is

```go
// gocb kv_pool_size - number of pipelines per node. Initialized on GetGoCBConnString
```

so it seems right to set it in here?

File: base/dcp_feed_type.go
Comment (markspolakovs) [reply base/dcp_feed_type.go:245]: It's not - the first set of functions use `StartDCPFeed`/`CouchbasePartitions`/`CouchbasePartitionSeqs`/`CouchbaseStats`, the second use `StartGocbcoreDCPFeed`/`CBPartitions`/`CBPartitionSeqs`/`CBStats`.

File: base/dcp_test.go
Comment (markspolakovs) [reply base/dcp_test.go]: We already have the capability - I've filed CBG-2216 to look into running it on a regular basis.

File: base/dcp_dest.go
Comment (bbrks) [inline base/dcp_dest.go]: I guess this is not the case any more with the gocb DCP 6.x compat - but if it were, would it mean an if/else around a capability and just setting `feedType` to one or the other?

File: base/dcp_feed_type.go
Comment (bbrks) [inline base/dcp_feed_type.go]: I guess this can be renamed to something like `cbgtGlobalsLock` now?

File: base/dcp_feed_type.go
Comment (bbrks) [inline base/dcp_feed_type.go]: We lock when we write to this map, but not when we read from it here. Should we be doing?

File: base/dcp_feed_type.go
Comment (bbrks) [inline base/dcp_feed_type.go:22]: What context do we only have a bucket name in?

This may not work if we have >1 database on a bucket and they're using different credentials/scopes.

The case I'm thinking of in particular is this:
- `db1` `bucket1` `scope1` `[col1,...]`
- `db2` `bucket1` `scope2` `[col1,...]`

Which could be a supported configuration - there's still only a single scope per db.

File: base/dcp_feed_type.go
Comment (bbrks) [inline base/dcp_feed_type.go:151]: Just for my own benefit - does GoCB/CBGT not allow streaming >1 scope?

**Review by bbrks — CHANGES_REQUESTED** [review]
Some initial comments

File: base/dcp_dest.go
Comment (markspolakovs) [reply base/dcp_dest.go]: Effectively - though given this appears to work just fine on 6.x that comment can likely be removed.

File: base/dcp_feed_type.go
Comment (markspolakovs) [reply base/dcp_feed_type.go]: Good catch

File: base/dcp_feed_type.go
Comment (markspolakovs) [reply base/dcp_feed_type.go:22]: `cbgt.GetPoolsDefaultForBucket` has signature `func(server, bucket string, scopes bool) ([]byte, error)`.

File: base/dcp_feed_type.go
Comment (markspolakovs) [reply base/dcp_feed_type.go:151]: I believe so:

https://github.com/couchbase/cbgt/blob/91cc2647a6b9cb775e476319ab52b5baf79d85a0/feed_dcp.go#L84-L88

File: base/dcp_feed_type.go
Comment (markspolakovs) [reply base/dcp_feed_type.go:22]: As discussed, I'll start the conversation with the cbgt maintainers about changing this, but let's not block this PR on that.

File: base/dcp_sharded.go
Comment (bbrks) [inline base/dcp_sharded.go]: Have you already seen `getRootCAs()` ? It looks like it's mostly doing the same thing as these load/system cert pool utils.

File: base/x509_utils.go
Comment (bbrks) [inline base/x509_utils.go]: SystemCertPool support for Windows was only added in 1.18, which I think means we may have to handle this slightly more gracefully?

In the other x509 util, we create an empty cert pool and let it fail when connecting if we error.

**Review by bbrks — CHANGES_REQUESTED** [review]
Another couple of comments but I think this PR now seems mostly there

**Review by bbrks — APPROVED** [review]

**Pull request merged**
